Add gRPC support to query-tee (#2683)
grafana/mimir: docs/sources/operators-guide/tools/query-tee.md

## Configure the query-tee

The query-tee requires the endpoints of the backend Grafana Mimir clusters.
You can configure the backend endpoints by setting the `-backend.endpoints` flag to a comma-separated list endpoints:

- HTTP endpoints: via HTTP or HTTPS URLs. Example: `http://query-frontend:80`.
- gRPC endpoints: via gRPC URI scheme. Example: `dns:///query-frontend:9095`.

>   For each incoming request, the query-tee clones the request and sends it to each configured backend.

Reviewer: Otherwise it's under gRPC endpoints, but it actually applies both to HTTP and gRPC.
```suggestion

For each incoming request, the query-tee clones the request and sends it to each configured backend.
```

> **Note:** You can configure the query-tee proxy listening ports via the `-server.service-port` flag for the HTTP port and `server.grpc-service-port` flag for the gRPC port.

## How the query-tee works

This section describes how the query-tee tool works.

### API endpoints

Query-tee accepts two types of requests:

1. HTTP requests on the configured `-server.service-port` flag (default port 80)

> 1. [HTTP over gRPC](https://github.com/weaveworks/common/tree/master/httpgrpc) requests on the configured `server.grpc-service-port` flag (default port: 9095)

Reviewer: ```suggestion
1. [HTTP over gRPC](https://github.com/weaveworks/common/tree/master/httpgrpc) requests on the configured `-server.grpc-service-port` flag (default port: 9095)
```

## Configure the query-tee

The query-tee requires the endpoints of the backend Grafana Mimir clusters.

> You can configure the backend endpoints by setting the `-backend.endpoints` flag to a comma-separated list endpoints:
> 
> - HTTP endpoints: via HTTP or HTTPS URLs. Example: `http://query-frontend:80`.
> - gRPC endpoints: via gRPC URI scheme. Example: `dns:///query-frontend:9095`.

Reviewer: I don't think this is correct. We only support HTTP for the backend. In this PR we just added gRPC support as "input" protocol.

```suggestion
You can configure the backend endpoints by setting the `-backend.endpoints` flag to a comma-separated list of HTTP or HTTPS URLs.
```

# TYPE cortex_querytee_responses_compared_total counter
cortex_querytee_responses_compared_total{route="<route>",result="<success|fail>"}
```

### Ruler remote operational mode test

> To test ruler evaluations with query-tee, set the `-ruler.query-frontend.address` CLI flag or its respective YAML configuration parameter for the ruler with query-tee's gRPC address:

Reviewer: ```suggestion
When the ruler is configured with the [remote evaluation mode]({{< relref "../architecture/components/ruler/index.md" >}}) you can use the query-tee to compare rule evaluations too.
To test ruler evaluations with query-tee, set the `-ruler.query-frontend.address` CLI flag or its respective YAML configuration parameter for the ruler with query-tee's gRPC address:
```

```
ruler:
  query_frontend:
    address: "dns://query-tee:9095"
```

When the ruler evaluates a rule, the test flow is the following:

1. ruler sends gRPC request to query-tee

> 1. query-tee forwards the request to the configured ruler-query-frontends of the Mimir instances: `-backend.endpoints` flag

Reviewer: ```suggestion
1. query-tee forwards the request to the query-frontend backends configured setting the `-backend.endpoints` CLI flag
```

> 1. query-tee receives the responses from the Mimir instances and forwards the result (based on the preferred backend) to the ruler

Reviewer: ```suggestion
1. query-tee receives the response from the query-frontend and forwards the result (based on the preferred backend) to the ruler
```
